### PR TITLE
Use Consistent ClusterState throughout Snapshot API Calls

### PR DESCRIPTION
**Pull request by original-brownbear**
We shouldn't be using potentially changing versions of the cluster state
when answering a snapshot status API call by calling `SnapshotService#currentSnapshots` multiple times (each time using `ClusterService#state` under the hood) but instead pass down the state from the transport action.
Having these API behave more in a more deterministic way will make it easier to use them once parallel repository operations
are introduced.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -1222,7 +1226,8 @@ public void deleteSnapshot(final String repositoryName, final String snapshotNam
             // if nothing found by the same name, then look in the cluster state for current in progress snapshots
             long repoGenId = repositoryData.getGenId();
             if (matchedEntry.isPresent() == false) {
-                Optional<SnapshotsInProgress.Entry> matchedInProgress = currentSnapshots(repositoryName, Collections.emptyList()).stream()
+                Optional<SnapshotsInProgress.Entry> matchedInProgress = currentSnapshots(
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: Making the fact that we're using a potentially racy CS here, leading to some needless snapshot delete failures, obvious was the main motivation behind this change actually (a bigger change to deletes making use of this cleanup is coming in https://github.com/elastic/elasticsearch/pull/51463).

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java [hunk 2]
```diff
@@ -103,14 +103,14 @@ protected void masterOperation(Task task, final GetSnapshotsRequest request, fin
...
-                                                () -> getMultipleReposSnapshotInfo(response.repositories(), request.snapshots(),
+                                                () -> getMultipleReposSnapshotInfo(state, response.repositories(), request.snapshots(),
                                                         request.ignoreUnavailable(), request.verbose(), listener)),
                                 listener::onFailure),
                         GetRepositoriesResponse::new));
     }
 
-    private void getMultipleReposSnapshotInfo(List<RepositoryMetaData> repos, String[] snapshots, boolean ignoreUnavailable,
-                                              boolean verbose, ActionListener<GetSnapshotsResponse> listener) {
+    private void getMultipleReposSnapshotInfo(ClusterState state, List<RepositoryMetaData> repos, String[] snapshots,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: Should we just pass down `SnapshotsInProgress`? Makes it a bit clearer what the dependencies are

**Review by tlrx — APPROVED** [review]
LGTM, I'm happy to see this change. Yannick's comment about `SnapshotsInProgress` should be addressed before merging if possible (does not seem to have an impact on #51463)

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: Sure sounds good :) => f2efe16

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1 (unrelated but interesting failure in S3 REST mock tests ... link for my reference: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/14597/testReport/junit/org.elasticsearch.repositories.s3/S3BlobContainerRetriesTests/testReadBlobWithRetries/ )

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java [hunk 3]
```diff
@@ -133,16 +135,16 @@ private void getMultipleReposSnapshotInfo(List<RepositoryMetaData> repos, String
...
                     }
-                }), wrappedListener -> getSingleRepoSnapshotInfo(repoName, snapshots, ignoreUnavailable, verbose,
+                }), wrappedListener -> getSingleRepoSnapshotInfo(snapshotsInProgress, repoName, snapshots, ignoreUnavailable, verbose,
                     ActionListener.map(wrappedListener, snInfos -> GetSnapshotsResponse.Response.snapshots(repoName, snInfos)))));
         }
     }
 
-    private void getSingleRepoSnapshotInfo(String repo, String[] snapshots, boolean ignoreUnavailable, boolean verbose,
-                                           ActionListener<List<SnapshotInfo>> listener) {
+    private void getSingleRepoSnapshotInfo(SnapshotsInProgress snapshotsInProgress, String repo, String[] snapshots,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: Can you add `@Nullable` to the methods that pass `SnapshotsInProgress`, or alternatively, pass an empty instance?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: Sure added the `@Nullable`s in 2bfe8f5 :)

**Review by ywelsch — APPROVED** [review]

**Comment by original-brownbear** [comment]
Thanks Tanguy + Yannick!

**Pull request merged**